### PR TITLE
Smithy codegen Client Refactor

### PR DESCRIPTION
**Pull request by AllanFly120**
A demo of refactored service client.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by AllanFly120** [comment]
One more thought on the config: Now I put all the config in the [config-resolver package](https://github.com/AllanFly120/aws-sdk-js-v3/blob/smithy-codegen-client-refactor/packages/config-resolver/src/components.ts). I think I would make more sense if I put `AuthConfig` to `signingMiddleware` package, put `retryConfig` to `retryMiddleware` package, put `protocolConfig` into httpProtocol package. This make codebase more modularized and show the relationship between middleware and its config

**Comment by AllanFly120** [comment]
I added a (de)serializer utils function bag to the interface. Check this commit: https://github.com/aws/aws-sdk-js-v3/pull/384/commits/72a02d9acee3c17755d669c61c28c87cc99231f7. I make the utils interface general on purpose because (De)Serializer interfaces are suppose to be share with non-AWS services, so no pre-assumption there.

**Comment by AllanFly120** [comment]
We need to add browser client smoke test scripts to this demo client also.

File: packages/stream-collector-node/src/index.ts [hunk 1]
```diff
@@ -2,9 +2,9 @@ import { Readable } from "stream";
...
 import { Collector } from "./collector";
 
-export const streamCollector: StreamCollector<
-  Readable
-> = function streamCollector(stream): Promise<Uint8Array> {
+export const streamCollector: StreamCollector = function streamCollector(
+  stream: Readable
+): Promise<Uint8Array> {
   return new Promise((resolve, reject) => {
     const collector = new Collector();
```
Comment (trivikr) [inline packages/stream-collector-node/src/index.ts]: nit: Convert to arrow function?

```typescript
export const streamCollector: StreamCollector = (
  stream: Readable
): Promise<Uint8Array> =>
  new Promise((resolve, reject) => {
    /* Code inside promise */
  });

```

File: packages/config-resolver/src/components.ts [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+  }
+  export interface Resolved {
+    maxRetries: number;
+    retryDecider?: RetryDecider;
+    delayDecider?: DelayDecider;
+  }
+  export function resolve<T>(input: T & Input): T & Resolved {
+    return {
+      ...input,
+      maxRetries: input.maxRetries === undefined ? 3 : input.maxRetries
```
[the remark this answers is not in the document]
Comment (AllanFly120) [reply packages/config-resolver/src/components.ts]: `maxRetries` can be set to 0.

File: clients/node/client-rds-data-node/RDSDataClient.ts [hunk 3]
```diff
@@ -23,32 +22,23 @@ const ServiceMetadata = {
...
+    const intermediaConfig_0 = ProtocolConfig.resolve({
+      ...RDSRuntimeConfiguration,
+      ...configuration
+    });
+    super(intermediaConfig_0);
+    let intermediaConfig_1 = RegionConfiguration.resolve(intermediaConfig_0);
+    let intermediaConfig_2 = AwsAuthConfiguration.resolve(intermediaConfig_1);
+    let intermediaConfig_3 = EndpointsConfig.resolve(intermediaConfig_2);
+    let intermediaConfig_4 = RetryConfig.resolve(intermediaConfig_3);
+    this.config = intermediaConfig_4;
```
[the remark this answers is not in the document]
Comment (AllanFly120) [reply clients/node/client-rds-data-node/RDSDataClient.ts]: This manner is stupidly easy for codegen. The advice looks normal but might be hard to codegen.

File: packages/stream-collector-node/src/index.ts (hunk 1 above)
Comment (srchase) [reply packages/stream-collector-node/src/index.ts]: fixed in bed5267da370e90567c7969fbb18beee281c5edf

File: clients/node/client-rds-data-node/protocol/AwsRestJson1_1.ts [hunk 4]
```diff
@@ -256,42 +264,93 @@ function columnMetadataListAwsRestJson1_1Deserialize(
...
   let list: Array<ColumnMetadata> = [];
-  for (let ColumnMetadata of input) {
-    list.push(columnMetadataAwsRestJson1_1Deserialize(ColumnMetadata));
+  if (input !== undefined) {
+    for (let ColumnMetadata of input) {
+      list.push(columnMetadataAwsRestJson1_1Deserialize(ColumnMetadata));
+    }
   }
   return list;
 }
```
[the remark this answers is not in the document]
Comment (srchase) [reply clients/node/client-rds-data-node/protocol/AwsRestJson1_1.ts]: fixed in bed5267da370e90567c7969fbb18beee281c5edf

File: clients/node/client-rds-data-node/protocol/AwsRestJson1_1.ts [hunk 5]
```diff
@@ -256,42 +264,93 @@ function columnMetadataListAwsRestJson1_1Deserialize(
...
-function parseBody(streamBody: any): any {
-  __aws_sdk_stream_collector_node.streamCollector(streamBody).then(body => {
-    return __aws_sdk_util_utf8_node.toUtf8(body);
+function deserializeMetadata(output: HttpResponse): ResponseMetadata {
+  return {
+    httpStatusCode: output.statusCode,
+    httpHeaders: output.headers,
+    requestId: output.headers["x-amzn-requestid"]
+  };
+}
```
[the remark this answers is not in the document]
Comment (srchase) [reply clients/node/client-rds-data-node/protocol/AwsRestJson1_1.ts]: fixed in bed5267da370e90567c7969fbb18beee281c5edf

File: packages/smithy-client/src/command.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import { MiddlewareStack } from "@aws-sdk/middleware-stack";
+import {
+  InjectableMiddleware,
+  HandlerOptions as InjectOptions
+} from "@aws-sdk/types";
+
+export class Command<InputType extends object, OutputType extends object> {
```
[the remark this answers is not in the document]
Comment (AllanFly120) [reply packages/smithy-client/src/command.ts]: Resolved by changing `SmithyClient` to `Client`

**Pull request merged**